README.md
# websaver
A Mac screen saver that shows a single webpage

Builds in XCode. Output is the directory WebSaver.saver which Finder sees as a screensaver file.


Changes.txt
2023/04/21 v2.8
    - Updated for macOS 10.13

2015/07/01 v2.7
    - WebGL support
    - Misc xcode cleanup

2012/09/12 v2.6
    - Params passing for multimonitor

2010/02/02 v2.5
	- Turned on Garbage Collection to support 64bit 
	  (http://lists.apple.com/archives/cocoa-dev/2009/Oct/msg00792.html)

2010/02/01 v2.4
	- Patched UniMotion to support 64bit IOKit

2009/09/06 v2.3
	- Rebuild for Snow Leopard 10.6

2009/05/26 v2.2
    - Repackaged and rebuilt for Google Code
	- No feature/bug changes
	
2008/03/21 v2.1
    - Added sudden motion sensor support on suitable hardware (thanks to the 
	  UniMotion http://sourceforge.net/projects/unimotion library).
	  The angle of the machine will be sent as (up, down, left, right) cursor
	  input to the loaded page. This might be usefull for scrolling pages 
	  without waking from screenaver.

2008/02/26 v2.0
    - Bumped major version number for initial public release


unimotion.h
/*
 *  UniMotion - Unified Motion detection for Apple portables.
 *
 *  Copyright (c) 2006 Lincoln Ramsay. All rights reserved.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License version 2.1 as published by the Free Software Foundation.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation Inc. 59 Temple Place, Suite 330, Boston MA 02111-1307 USA
 */
#ifndef UNIMOTION_H
#define UNIMOTION_H

#include <stdint.h>

#ifdef __cplusplus
extern "C" {
#endif

// The various SMS hardware that unimotion supports
enum sms_hardware {
    unknown = 0,
    powerbook = 1,
    ibook = 2,
    highrespb = 3,
    macbookpro = 4
};

// prototypes for the functions in unimotion.c

// returns the value of SMS hardware present or unknown if no hardware is
// detected
int detect_sms(void);

// use the value returned from detect_sms as the type
// don't call read_sms(detect_sms()...) as this will do extra work
// if you can't save the type between calls pass 0 as type to avoid extra work

// all functions return 1 on success and 0 on failure
// they modify x, y and z if they are not 0


//
// 0.3 functions
//

// raw, unmodified values
int read_sms_raw(int type, int *x, int *y, int *z);

// "calibrated" values (same as raw if no calibration data exists)
int read_sms(int type, int *x, int *y, int *z);
// real (1.0 = 1G) values (requires calibration data)
// note that this is the preferred API as it need not change with new machines
// if no "scale" calibration data exists defaults will be used based on the
// machine type
int read_sms_real(int type, double *x, double *y, double *z);


//
// 0.4 functions
//

// raw SMS data (useful for debugging)
// note that endian issues make reading the raw bytes non-trivial
// returns 0 or a pointer that should be released with free() when you are
// finished with it
// sets *size to the size of the structure returned if size is not 0
uint8_t *read_sms_raw_bytes(int type, int *size);

// scaled values, like real but easier to handle
// this reverses the backwards polarity of x and increases the range of the
// older machines to match the MacBook [Pro] sensor
int read_sms_scaled(int type, int *x, int *y, int *z);

#ifdef __cplusplus
}
#endif

#endif



WebSaverView.h
//
//  WebSaverView.h
//
//    WebSaver - A web screen-saver for MacOS-X
//    Copyright (C) 2008 Gavin Brock http://brock-family.org/gavin
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#import <ScreenSaver/ScreenSaver.h>
#import <WebKit/WebKit.h>
#include "unimotion.h"


@interface WebSaverView : ScreenSaverView 
{
    WKWebView *webView;
	NSDate *lastLoad;
	
	// Motion Sensor Related
	int sms_type, avgx, avgy, avgz;
    
	IBOutlet id configSheet;

	IBOutlet id versionLabel;

	IBOutlet NSTextField *saverURL;
	NSString *saverURLString;
	
	IBOutlet id enableReload;
	bool enableReloadBool;

	IBOutlet NSPopUpButton *reloadTime;
	long reloadTimeFloat;
	
	IBOutlet id enableSMS;
	bool enableSMSBool;

    IBOutlet id enableMultiMonitor;
	bool enableMultiMonitorBool;

}
@end


WebSaverView.m
//
//  WebSaverView.m
//
//    WebSaver - A web screen-saver for MacOS-X
//    Copyright (C) 2008 Gavin Brock http://brock-family.org/gavin
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// Cheesy Debugging toggle
//#define DebugLog //
#define DebugLog NSLog


#import "WebSaverView.h"

static NSString * const MyModuleName = @"org.brock-family.WebSaver";
static NSString * upArrow, *downArrow, *leftArrow, *rightArrow;


@implementation WebSaverView

- (void)webView:(WKWebView *)sender didStartProvisionalLoadForFrame:(WKWebView *)frame
{
	DebugLog(@"webView:didStartProvisionalLoadForFrame: %@", frame);
}

- (void)webView:(WKWebView *)sender didCommitLoadForFrame:(WKWebView *)frame
{
	DebugLog(@"webView:didCommitLoadForFrame");
}

- (void)webView:(WKWebView *)sender didFailLoadWithError:(NSError *)error forFrame:(WKWebView *)frame
{
	DebugLog(@"webView:didFailLoadWithError: %@", error);
}

- (void)webView:(WKWebView *)sender didFailProvisionalLoadWithError:(NSError *)error forFrame:(WKWebView *)frame
{
	DebugLog(@"webView:didFailProvisionalLoadWithError: %@", error);
}

- (void)webView:(WKWebView *)sender didFinishLoadForFrame:(WKWebView *)frame
{
	DebugLog(@"webView:didFinishLoadForFrame");
}

- (void)webView:(WKWebView *)sender didReceiveIcon:(NSImage *)image forFrame:(WKWebView *)frame
{
	DebugLog(@"webView:didReceiveIcon");
}

- (void)webView:(WKWebView *)sender serverRedirectedForDataSource:(WKWebView *)frame
{
	DebugLog(@"webView:serverRedirectedForDataSource");
}

- (void)webView:(WKWebView *)sender didReceiveTitle:(NSString *)title forFrame:(WKWebView *)frame
{
	NSLog(@"webView:didReceiveTitle: %@", title);
}



- (id)initWithFrame:(NSRect)frame isPreview:(BOOL)isPreview
{
        
    DebugLog(@"webView:initWithFrame isPreview:%d", isPreview);
    self = [super initWithFrame:frame isPreview:isPreview];
    if (self) {
	
		sms_type = detect_sms();
		if (sms_type) {
			unichar ua = NSUpArrowFunctionKey;
			unichar da = NSDownArrowFunctionKey;
			unichar la = NSLeftArrowFunctionKey;
			unichar ra = NSRightArrowFunctionKey;
			upArrow    = [[NSString alloc] initWithCharacters:&ua length:1];
			downArrow  = [[NSString alloc] initWithCharacters:&da length:1];
			leftArrow  = [[NSString alloc] initWithCharacters:&la length:1];
			rightArrow = [[NSString alloc] initWithCharacters:&ra length:1];	
		}

		ScreenSaverDefaults *defaults;
		defaults = [ScreenSaverDefaults defaultsForModuleWithName:MyModuleName];
		[defaults registerDefaults:[NSDictionary dictionaryWithObjectsAndKeys: 
			@"http://www.apple.com/",  @"URL",
			@"NO",                     @"EnableReload",
			@"300",                    @"ReloadTime",
			@"NO",                     @"EnableSMS",
            @"NO",                     @"EnableMultiMonitor",
			nil ]];
		
		[self setAnimationTimeInterval:0.5];
		
		saverURLString = [defaults stringForKey:@"URL"];
		DebugLog(@"URL: %@", saverURLString);
		
		enableReloadBool = [defaults boolForKey:@"EnableReload"];
		DebugLog(@"Will Reload: %d", enableReloadBool);
		
		reloadTimeFloat = [defaults floatForKey:@"ReloadTime"];
		DebugLog(@"Reload Time: %ld", reloadTimeFloat);

		enableSMSBool = [defaults boolForKey:@"EnableSMS"];
		DebugLog(@"Will use SMS: %d", enableSMSBool);
        
		enableMultiMonitorBool = [defaults boolForKey:@"EnableMultiMonitor"];
		DebugLog(@"Will use MultiMonitor: %d", enableMultiMonitorBool);
        
		webView = [[WKWebView alloc] initWithFrame:frame];
//		[webView setDrawsBackground:NO];
        [webView setCustomUserAgent:@"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8) AppleWebKit/536.25 (KHTML, like Gecko) Version/6.0 Safari/536.25"];

        if (isPreview) {
			[self scaleUnitSquareToSize:NSMakeSize( 0.25, 0.25 )];
		}

		[self addSubview:webView];
	}

    return self;
}

- (void)setFrame:(NSRect)frameRect
{
	DebugLog(@"NSView:setFrame");
	//DebugLog(@"frameRect %d,%d %dx%d\n", frameRect.origin.x, frameRect.origin.y, frameRect.size.width, frameRect.size.height);
	[super setFrame:frameRect];
	[webView setFrame:frameRect];	
	[webView setFrameSize:[webView convertSize:frameRect.size fromView:nil]];
}


- (void)startAnimation
{ 
    NSString *url;
    NSScreen *screen;
    int srand_time = 15;
    
	DebugLog(@"webView:startAnimation");
    [super startAnimation];
    
	// Calibrate 'natural' positon with 20 readings
	if (enableSMSBool && sms_type) {
		int loop;	
		for(loop = 0; loop < 20; loop++) {
			int x, y, z;
			if (read_sms_scaled(sms_type, &x, &y, &z)) {
				avgx = avgx + x;
				avgy = avgy + y;
				avgz = avgz + z;
			}
		}
		avgx = avgx / 20;
		avgy = avgy / 20;
		avgz = avgz / 20; 
		DebugLog(@"SMS Average - x = %d, y = %d, z = %d", avgx, avgy, avgz);
	}

    if (enableMultiMonitorBool) {
        screen = [ [webView window] screen ];
        DebugLog(@"Screen %@", [screen description]);
        url = [NSString stringWithFormat: @"%@?x=%.0f&y=%.0f&w=%.0f&h=%.0f&screen=%i&srand=%i", saverURLString,
               [screen frame].origin.x,   [screen frame].origin.y,
               [screen frame].size.width, [screen frame].size.height,
               (int)[[NSScreen screens] indexOfObject: screen],
               ((int)time(0)+(srand_time/2)/srand_time) ];
        DebugLog(@"URL with query: %@", url );
    } else {
        url = saverURLString;
    }
        
    [self loadUrl];

}

- (void)stopAnimation
{
	DebugLog(@"webView:stopAnimation");
	
    [webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"about:"]]];

    [super stopAnimation];
}

- (void)doKeyUp:(NSTimer*)theTimer {
	//DebugLog(@"doKeyUp");
    [[NSApplication sharedApplication] sendEvent:[NSEvent keyEventWithType:NSEventTypeKeyUp
		location:      NSMakePoint(1,1)
		modifierFlags: 0
		timestamp:     [[NSDate date] timeIntervalSinceReferenceDate]
		windowNumber:  [[self window] windowNumber] 
		context:       [NSGraphicsContext currentContext] 
		characters:    [theTimer userInfo]
		charactersIgnoringModifiers:[theTimer userInfo]
		isARepeat:     NO 
		keyCode:       0]];
}


- (void)doKeyDown:(NSTimer*)theTimer {
	//DebugLog(@"doKeyDown");
    [[NSApplication sharedApplication] sendEvent:[NSEvent keyEventWithType:NSEventTypeKeyDown
		location:      NSMakePoint(1,1)
		modifierFlags: 0
		timestamp:     [[NSDate date] timeIntervalSinceReferenceDate]
		windowNumber:  [[self window] windowNumber] 
		context:       [NSGraphicsContext currentContext] 
		characters:    [theTimer userInfo]
		charactersIgnoringModifiers:[theTimer userInfo]
		isARepeat:     NO 
		keyCode:      0]];
}


- (void)animateOneFrame
{
	// TOO VERBOSE // DebugLog(@"webView:animateOneFrame");
	if (enableReloadBool) {
		// DebugLog(@"time %f", [lastLoad timeIntervalSinceNow]);
		if (reloadTimeFloat + [lastLoad timeIntervalSinceNow] < 0) {
            [webView reload];
			lastLoad = [[NSDate alloc] init];
			DebugLog(@"reloaded %@",[lastLoad description]);
		}
	}
		
	if (enableSMSBool && sms_type) {
		int x,y,z;
		if (read_sms_scaled(sms_type, &x, &y, &z)) {
			//DebugLog(@"average - x = %d, y = %d, z = %d     Current Reading:  x = %d, y = %d, z = %d", avgx, avgy, avgz, x, y, z);
	
			int cx = avgx - x;
			int cy = avgy - y;
		
			if (abs(cx) > abs(cy)) {
				if(cx > 30) {
					DebugLog(@"SMS Send Right Cursor");
					[NSTimer scheduledTimerWithTimeInterval:(NSTimeInterval)0 target:self selector:@selector(doKeyDown:) userInfo:rightArrow repeats:NO];
					[NSTimer scheduledTimerWithTimeInterval:(NSTimeInterval).05 target:self selector:@selector(doKeyUp:) userInfo:rightArrow repeats:NO];
				} else if(cx < -30) {
					DebugLog(@"SMS Send Left Cursor");
					[NSTimer scheduledTimerWithTimeInterval:(NSTimeInterval)0 target:self selector:@selector(doKeyDown:) userInfo:leftArrow repeats:NO];
					[NSTimer scheduledTimerWithTimeInterval:(NSTimeInterval).05 target:self selector:@selector(doKeyUp:) userInfo:leftArrow repeats:NO];
				}
			} else {
				if(cy > 30) {
					DebugLog(@"SMS Send Up Cursor");
					[NSTimer scheduledTimerWithTimeInterval:(NSTimeInterval)0 target:self selector:@selector(doKeyDown:) userInfo:upArrow repeats:NO];
					[NSTimer scheduledTimerWithTimeInterval:(NSTimeInterval).05 target:self selector:@selector(doKeyUp:) userInfo:upArrow repeats:NO];
				} else if(cy < -30) {
					DebugLog(@"SMS Send Down Cursor");
					[NSTimer scheduledTimerWithTimeInterval:(NSTimeInterval)0 target:self selector:@selector(doKeyDown:) userInfo:downArrow repeats:NO];
					[NSTimer scheduledTimerWithTimeInterval:(NSTimeInterval).5 target:self selector:@selector(doKeyUp:) userInfo:downArrow repeats:NO];
				}	
			}
		}
	}
    return;
}



- (BOOL)hasConfigureSheet
{
	DebugLog(@"webView:hasConfigureSheet");
    return YES;
}

- (NSWindow*)configureSheet
{
	DebugLog(@"webView:configureSheet");
	
	if (!configSheet) {
// This should remover the 10.8 waring - but then the sheet does not open
// if (![[NSBundle mainBundle] loadNibNamed:@"ConfigureSheet" owner:self topLevelObjects:nil]) {
        if (![NSBundle loadNibNamed:@"ConfigureSheet" owner:self]) {
			NSLog( @"Failed to load configure sheet." );
			NSBeep(); 
		} 
	}

	NSDictionary *infoDictionary;
	infoDictionary = [[NSBundle bundleForClass:[self class]] infoDictionary];
	NSString *versionString;
	versionString = [infoDictionary objectForKey:@"CFBundleVersion"];

	[versionLabel setStringValue:versionString];
	[saverURL setStringValue:saverURLString];
	[enableReload setState:enableReloadBool];
	[reloadTime selectItemWithTag:reloadTimeFloat];
	[enableSMS setState:enableSMSBool];
	[enableMultiMonitor setState:enableMultiMonitorBool];
	[reloadTime setEnabled:[enableReload state]];
	[enableSMS setEnabled:sms_type ? true : false];

	return configSheet;
}


- (IBAction)enableClick:(id)sender
{
	DebugLog(@"IBAction:enableClick");
	[reloadTime setEnabled:[enableReload state]];
}


- (IBAction)cancelClick:(id)sender
{
	DebugLog(@"IBAction:cancelClick");
	[[NSApplication sharedApplication] endSheet:configSheet];
}


- (IBAction)okClick:(id)sender
{
	DebugLog(@"IBAction:okClick");

	// Check what the user set
	saverURLString = [saverURL stringValue];
	NSLog(@"Set URL: %@", saverURLString);
	enableReloadBool = [enableReload state];
	NSLog(@"Set Will Reload: %d", enableReloadBool);
	reloadTimeFloat = [[reloadTime selectedItem] tag];
	NSLog(@"Set Reload Time: %ld", reloadTimeFloat);
	enableSMSBool = [enableSMS state];
	NSLog(@"Set Use SMS: %d", enableSMSBool);
	enableMultiMonitorBool = [enableMultiMonitor state];
	NSLog(@"Set Use Multi Monitor: %d", enableMultiMonitorBool);


	// Save the Settings
	ScreenSaverDefaults *defaults;
	defaults = [ScreenSaverDefaults defaultsForModuleWithName:MyModuleName];
	[defaults setObject:saverURLString  forKey:@"URL"];
	[defaults setBool:enableReloadBool  forKey:@"EnableReload"];
	[defaults setFloat:reloadTimeFloat  forKey:@"ReloadTime"];
	[defaults setBool:enableSMSBool     forKey:@"EnableSMS"];
	[defaults setBool:enableMultiMonitorBool forKey:@"EnableMultiMonitor"];
	[defaults synchronize];
	
    [self loadUrl];
    
	// Close the window
	[[NSApplication sharedApplication] endSheet:configSheet];
}
- (void)loadUrl
{
    // Reload the page and reset load time
    NSMutableURLRequest *request =
    [NSMutableURLRequest requestWithURL:[NSURL URLWithString:saverURLString]];
    
    //TODO - add this to menu
    //[request setHTTPMethod:@"POST"];
    
    [webView loadRequest:request];
    lastLoad = [[NSDate alloc] init];
    DebugLog(@"reloaded %@",[lastLoad description]);
}


@end


unimotion.c
 /*
 *  UniMotion - Unified Motion detection for Apple portables.
 *
 *  Copyright (c) 2006 Lincoln Ramsay. All rights reserved.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License version 2.1 as published by the Free Software Foundation.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation Inc. 59 Temple Place, Suite 330, Boston MA 02111-1307 USA
 */

/*
 * HISTORY of Motion
 * Written by Christian Klein
 * Modified for iBook compatibility by Pall Thayer
 * Modified for Hi Res Powerbook compatibility by Pall Thayer
 * Modified for MacBook Pro compatibility by Randy Green
 * Disparate forks unified into UniMotion by Lincoln Ramsay
 */

// This license applies to the portions created by Cristian Klein.
/* motion.c
 *
 * a little program to display the coords returned by
 * the powerbook motion sensor
 *
 * A fine piece of c0de, brought to you by
 *
 *               ---===---
 * *** teenage mutant ninja hero coders ***
 *               ---===---
 *
 * All of the software included is copyrighted by Christian Klein <chris@5711.org>.
 *
 * Copyright 2005 Christian Klein. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author must not be used to endorse or promote
 *    products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include "unimotion.h"
#include <IOKit/IOKitLib.h>
#include <CoreFoundation/CoreFoundation.h>
#include <string.h>
#include <stdint.h>

enum data_type {
    PB_IB,
    MBP
};

struct pb_ib_data {
    int8_t x;
    int8_t y;
    int8_t z;
    int8_t pad[57];
};

struct mbp_data {
    int16_t x;
    int16_t y;
    int16_t z;
    int8_t pad[34];
};

union motion_data {
    struct pb_ib_data pb_ib;
    struct mbp_data mbp;
};

struct default_real_values {
    double xreal;
    double yreal;
    double zreal;
};

static struct default_real_values drv[] = {
    {0.0,0.0,0.0},             // unknown
    {2.0e-2,2.0e-2,2.0e-2},    // powerbook
    {2.0e-2,2.0e-2,2.0e-2},    // ibook
    {-2.0e-2,-2.0e-2,-2.0e-2},  // highrespb
    {4.0e-3,4.0e-3,4.0e-3},    // macbookpro
};

struct default_scale_values {
    int xscale;
    int yscale;
    int zscale;
};

static struct default_scale_values dsv[] = {
    {0,0,0},    // unknown
    {5,5,5},    // powerbook
    {5,5,5},    // ibook
    {-5,-5,-5},  // highrespb
    {1,1,1},    // macbookpro
};

static int set_values(int type, int *kernFunc, char **servMatch, int *dataType)
{
    switch ( type ) {
        case powerbook:
            *kernFunc = 21;
            *servMatch = "IOI2CMotionSensor";
            *dataType = PB_IB;
            break;
        case ibook:
            *kernFunc = 21;
            *servMatch = "IOI2CMotionSensor";
            *dataType = PB_IB;
            break;
        case highrespb:
            *kernFunc = 21;
            *servMatch = "PMUMotionSensor";
            *dataType = PB_IB;
            break;
        case macbookpro:
            *kernFunc = 5;
            *servMatch = "SMCMotionSensor";
            *dataType = MBP;
            break;
        default:
            return 0;
    }

    return 1;
}

static int probe_sms(int kernFunc, char *servMatch, int dataType, void *data)
{
    kern_return_t result;
    mach_port_t masterPort;
    io_iterator_t iterator;
    io_object_t aDevice;
    io_connect_t  dataPort;

    IOItemCount structureInputSize;
    size_t structureOutputSize;

    union motion_data inputStructure;
    union motion_data *outputStructure;

    outputStructure = (union motion_data *)data;

    result = IOMasterPort(MACH_PORT_NULL, &masterPort);

    CFMutableDictionaryRef matchingDictionary = IOServiceMatching(servMatch);

    result = IOServiceGetMatchingServices(masterPort, matchingDictionary, &iterator);

    if (result != KERN_SUCCESS) {
        //fputs("IOServiceGetMatchingServices returned error.\n", stderr);
        return 0;
    }

    aDevice = IOIteratorNext(iterator);
    IOObjectRelease(iterator);

    if (aDevice == 0) {
        //fputs("No motion sensor available\n", stderr);
        return 0;
    }

    result = IOServiceOpen(aDevice, mach_task_self(), 0, &dataPort);
    IOObjectRelease(aDevice);

    if (result != KERN_SUCCESS) {
        //fputs("Could not open motion sensor device\n", stderr);
        return 0;
    }

    switch ( dataType ) {
        case PB_IB:
            structureInputSize = sizeof(struct pb_ib_data);
            structureOutputSize = sizeof(struct pb_ib_data);
            break;
        case MBP:
            structureInputSize = sizeof(struct mbp_data);
            structureOutputSize = sizeof(struct mbp_data);
            break;
        default:
            return 0;
    }

    memset(&inputStructure, 0, sizeof(union motion_data));
    memset(outputStructure, 0, sizeof(union motion_data));

#if !defined(__LP64__)
    // Check if Mac OS X 10.5 API is available...
    if (IOConnectCallMethod != NULL) {
        // ...and use it if it is.
#endif
		result = IOConnectCallStructMethod(dataPort, kernFunc, &inputStructure, structureInputSize,
				outputStructure, &structureOutputSize);
#if !defined(__LP64__)
    }
    else {
        // Otherwise fall back to older API.		
		result = IOConnectMethodStructureIStructureO(dataPort, kernFunc, structureInputSize,
			  &structureOutputSize, &inputStructure, outputStructure);
    }
#endif
		
	
    IOServiceClose(dataPort);

    if (result != KERN_SUCCESS) {
        //puts("no coords");
        return 0;
    }
    return 1;
}

int detect_sms(void)
{
    int kernFunc;
    char *servMatch;
    int dataType;
    union motion_data data;
    int i;

    for ( i = 1; ; i++ ) {
        if ( !set_values(i, &kernFunc, &servMatch, &dataType) )
            break;
        if ( probe_sms(kernFunc, servMatch, dataType, &data) )
            return i;
    }

    return unknown;
}

int read_sms_raw(int type, int *x, int *y, int *z)
{
    int kernFunc;
    char *servMatch;
    int dataType;
    union motion_data data;
    int t;

    t = type;
    if ( t == 0 ) {
        int i;

        for ( i = 1; ; i++ ) {
            if ( !set_values(i, &kernFunc, &servMatch, &dataType) )
                break;
            if ( probe_sms(kernFunc, servMatch, dataType, &data) ) {
                t = i;
                break;
            }
        }
    } else {
        if ( !set_values(type, &kernFunc, &servMatch, &dataType) )
            return 0;
        if ( !probe_sms(kernFunc, servMatch, dataType, &data) )
            return 0;
    }

    if ( t ) {
        switch ( dataType ) {
            case PB_IB:
                if ( x ) *x = data.pb_ib.x;
                if ( y ) *y = data.pb_ib.y;
                if ( z ) *z = data.pb_ib.z;
                break;
            case MBP:
                if ( x ) *x = data.mbp.x;
                if ( y ) *y = data.mbp.y;
                if ( z ) *z = data.mbp.z;
                break;
            default:
                return 0;
        }
        return 1;
    }
    return 0;
}

int read_sms(int type, int *x, int *y, int *z)
{
    int _x, _y, _z;
    long xoff, yoff, zoff;
    Boolean ok;
    int ret;

    ret = read_sms_raw(type, &_x, &_y, &_z);
    if ( !ret )
        return 0;

    static CFStringRef app = CFSTR("com.ramsayl.UniMotion");
    static CFStringRef xoffstr = CFSTR("x_offset");
    static CFStringRef yoffstr = CFSTR("y_offset");
    static CFStringRef zoffstr = CFSTR("z_offset");
    xoff = CFPreferencesGetAppIntegerValue(xoffstr, app, &ok);
    if ( ok ) _x += xoff;
    yoff = CFPreferencesGetAppIntegerValue(yoffstr, app, &ok);
    if ( ok ) _y += yoff;
    zoff = CFPreferencesGetAppIntegerValue(zoffstr, app, &ok);
    if ( ok ) _z += zoff;

    *x = _x;
    *y = _y;
    *z = _z;

    return ret;
}

static Boolean getPrefDouble(CFStringRef key, CFStringRef app, double *val)
{
    CFPropertyListRef prefRef;
    Boolean ok;
    double ret;

    ok = false;
    prefRef = CFPreferencesCopyAppValue(key, app);
    if ( prefRef ) {
        if ( CFGetTypeID(prefRef) == CFNumberGetTypeID() && CFNumberIsFloatType(prefRef) ) {
            ok = CFNumberGetValue(prefRef, kCFNumberDoubleType, &ret);
            if ( ok && val )
                *val = ret;
        }
        CFRelease(prefRef);
    }
    return ok;
}

int read_sms_real(int type, double *x, double *y, double *z)
{
    int _x, _y, _z;
    long xscale, yscale, zscale;
    double xreal, yreal, zreal;
    int ret;
    Boolean ok;

    // start with the "calibrated" (offset applied) value
    ret = read_sms(type, &_x, &_y, &_z);
    if ( !ret )
        return 0;

    static CFStringRef app = CFSTR("com.ramsayl.UniMotion");
    static CFStringRef xrealstr = CFSTR("x_real");
    static CFStringRef yrealstr = CFSTR("y_real");
    static CFStringRef zrealstr = CFSTR("z_real");
    static CFStringRef xscalestr = CFSTR("x_scale");
    static CFStringRef yscalestr = CFSTR("y_scale");
    static CFStringRef zscalestr = CFSTR("z_scale");

#define codeblock(x,_x,xreal,xscale) \
    ok = getPrefDouble(xreal##str, app, &xreal); \
    if ( ok ) { \
        *x = _x * xreal; /* use x_real if it's there */ \
    } else { \
        xscale = CFPreferencesGetAppIntegerValue(xscale##str, app, &ok); \
        if ( ok ) { \
            *x = _x / (double)xscale; /* try x_scale next */ \
        } else { \
            *x = _x * drv[type].xreal; /* fall back to "standard" values */ \
        } \
    }

    codeblock(x,_x,xreal,xscale);
    codeblock(y,_y,yreal,yscale);
    codeblock(z,_z,zreal,zscale);

#undef codeblock

    return 1;
}

uint8_t *read_sms_raw_bytes(int type, int *size)
{
    int kernFunc;
    char *servMatch;
    int dataType;
    uint8_t *data;
    int t;

    data = malloc(sizeof(union motion_data));
    if ( data == 0 )
        return 0;

    t = type;
    if ( t == 0 ) {
        int i;

        for ( i = 1; ; i++ ) {
            if ( !set_values(i, &kernFunc, &servMatch, &dataType) )
                break;
            if ( probe_sms(kernFunc, servMatch, dataType, data) ) {
                t = i;
                break;
            }
        }
    } else {
        if ( !set_values(type, &kernFunc, &servMatch, &dataType) )
            return 0;
        if ( !probe_sms(kernFunc, servMatch, dataType, data) )
            return 0;
    }

    if ( t ) {
        switch ( dataType ) {
            case PB_IB:
                if ( size )
                    *size = sizeof(struct pb_ib_data);
                break;
            case MBP:
                if ( size )
                    *size = sizeof(struct mbp_data);
                break;
            default:
                t = 0;
                break;
        }
        if ( t )
            return data;
    }
    free(data);
    if ( size )
        *size = 0;
    return 0;
}

int read_sms_scaled(int type, int *x, int *y, int *z)
{
    int _x, _y, _z;
    int ret;

    ret = read_sms(type, &_x, &_y, &_z);
    if ( !ret )
        return 0;

    *x = _x + dsv[type].xscale;
    *y = _y + dsv[type].yscale;
    *z = _z + dsv[type].zscale;

    return ret;
}

